Ios migration guide (#45)
MapsPeople/docsite: docs/01-getting-started/ios/v4/v4-migration-guide.mdx

Reviewer: This needs the content @fhansenMP has written somewhere…

---
title: Migrating from v3 to v4
hide_title: false
hide_table_of_contents: true
sidebar_position: 0
---
The iOS SDK for MapsIndoors has received a major update to version 4.0.0, which comes with improved interfaces and flexibility for developing your own map experience. The MapsIndoors SDK now additionally supports Mapbox, alongside some reworked and refactored features that simplify development and SDK behavior. This guide will cover changes to the SDK and how to use it to provide you with a guide on how to upgrade from V3 to V4. _If you have any questions concerning this document, or migrating to v4 in general, please contact MapsPeople technical support._

> MapsIndoors SDK Map Engine Flavors[​](https://docsite.mapsindoors.com/getting-started/android/v4/#mapsindoors-sdk-map-engine-flavors)

Reviewer: References Android Getting Started Guide – is this on purpose?

-------------------------------------------------------------------------------------------------------------------------------------

With the release of V4, the MapsIndoors SDK is released as multiple separate libraries depending on the Map Provider - Google Maps or Mapbox. You now import the SDK flavor of your choosing with either:

> ```swift

Reviewer: ```suggestion
```ruby
```

pod 'MapsIndoorsGoogleMaps', '~> 4.0.0'
```
or

> ```swift

Reviewer: ```suggestion
```ruby
```

pod 'MapsIndoorsMapbox', '~> 4.0.0'
```

> MapsIndoors Initialization[​](https://docsite.mapsindoors.com/getting-started/android/v4/#mapsindoors-initialization) & Usage

Reviewer: References Android Getting Started Guide – is this on purpose?

Reviewer 2: nope, error - fixing all urls to 'ios' instead of 'android'

In V3 you would start the SDK by providing a MapsIndoors API key and a Google API key, in order to start loading a solution.

```swift
MPMapsIndoors.provideAPIKey("YOUR_MAPSINDOORS_API_KEY", googleAPIKey: "YOUR_GOOGLE_API_KEY")
```

In order to wait until MapsIndoors had finished loading the solution, you would call `synchronizeContent` with a completion.

```swift
MPMapsIndoors.synchronizeContent { error in
  // The SDK has finished loading
}
```

In V3 you had no real way to close the MapsIndoors SDK, releasing system resources when no map is shown.

#### MapsIndoors Data

In v3, you would access MapsIndoors data via data providers or services, e.g:

```swift
let solution = try await MPSolutionProvider().solution()
                
let venues = try await MPVenueProvider().venues()

let locations1 = try await MPLocationsProvider().locations()
                
let locations2 = try await MPLocationService.sharedInstance().locations(using: MPQuery(), filter: MPFilter())
```

#### V4

In V4, we have improved the interface to initiate the SDK for smaller and safer implementations. In V4 you start loading a solution with a MapsIndoors API by:

```swift

> let error = await MPMapsIndoors.shared.load(apiKey: "YOUR_MAPSINDOORS_API_KEY")

Reviewer: ```suggestion
do {
    try await MPMapsIndoors.shared.load(apiKey: "YOUR_MAPSINDOORS_API_KEY")
} catch {
    // Do something with the thrown error
    print(error)
}
```

```

> Note that this is async/await and therefore requires an asynchronous context, and returns a potential loading error. You can therefore be sure on the next line, that the SDK has either successfully loaded, or failed.

Reviewer: ```suggestion
Note that this is async/await and therefore requires an asynchronous context, and potentially throws an error if there were problems during loading MapsIndoors data. You can therefore be sure on the next line, that the SDK has either successfully loaded, or failed.
```

```

MapControl Initialization[​](https://docsite.mapsindoors.com/getting-started/android/v4/#mapcontrol-initialization)
-------------------------------------------------------------------------------------------------------------------

MapControl instantiation and initialization are separate concepts. You create a new instance of `MPMapControl` and configure it with a map and view.

#### V3

To create a `MPMapControl` instance in V3, it was really straight forward.

```swift
let mapControl = MPMapControl(map: googleMapView)
```

#### V4

In V4, in order to support multiple map engines, we have changed how you create a `MPMapControl` instance. You now have to call a factory function on `MapsIndoors`, and provide an `MPMapConfig`.

```swift
if let mapControl = MPMapsIndoors.createMapControl(mapConfig: mapConfig) {
  // MapControl instance
}
```

An `MPMapConfig` is a configurable wrapper around either a Google Maps or Mapbox view. Depending on which flavor of the SDK you are using, either of the following approaches will derive a `MPMapConfig`, which you can use to create a `MPMapControl`:

Google:

```swift

> let mapConfig = MPMapConfig.init(gmsMapView: googleMapsView, googleApiKey: "AIza....")

Reviewer: ```suggestion
let mapConfig = MPMapConfig(gmsMapView: googleMapsView, googleApiKey: "AIza....")
```

```

Mapbox:

```swift

> let mapConfig = MPMapConfig.init(mapBoxView: mapboxMapView, accessToken: "mapbox api key")

Reviewer: ```suggestion
let mapConfig = MPMapConfig(mapBoxView: mapboxMapView, accessToken: "mapbox api key")
```

```swift
var rememberedRule: MPLocationDisplayRule?

// Modifying for highlight
let location = try await MPLocationsProvider().location(withId: "myLocationId")
if let displayRule = mapControl?.getEffectiveDisplayRule(for: location) {
    rememberedRule = displayRule.copy() as? MPLocationDisplayRule
    displayRule.showPolygon = true
    displayRule.polygonFillColor = .red
}

// Resetting
if let originalDisplayRule = rememberedRule {
    mapControl?.setDisplayRule(originalDisplayRule, for: location)
}
```

##### Editing built-in SDK display rules

You could access the built-in display rules either directly on your `MPMapControl` instance, or by name.

```swift
let selectionHighlight = mapControl?.locationHighlightDisplayRule
let blueDotDisplayRule = mapControl?.getDisplayRule(forTypeNamed:"my-location")
```

#### V4

##### Editing a display rule

```swift
if let location = MPMapsIndoors.shared.locationWith(locationId: "myLocationId") {

>     if let displayRule = MPMapsIndoors.shared.disPplayRuleFor(location: location) {

Reviewer: ```suggestion
    if let displayRule = MPMapsIndoors.shared.displayRuleFor(location: location) {
```
